// @backstage/plugin-search-backend-module-techdocs/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendCommon = require('@backstage/backend-common');
var backendPluginApi = require('@backstage/backend-plugin-api');
var backendTasks = require('@backstage/backend-tasks');
var alpha = require('@backstage/plugin-catalog-node/alpha');
var pluginSearchBackendModuleTechdocs = require('@backstage/plugin-search-backend-module-techdocs');
var alpha$1 = require('@backstage/plugin-search-backend-node/alpha');

const searchModuleTechDocsCollator = backendPluginApi.createBackendModule({
  moduleId: "techDocsCollator",
  pluginId: "search",
  register(env) {
    env.registerInit({
      deps: {
        config: backendPluginApi.coreServices.rootConfig,
        logger: backendPluginApi.coreServices.logger,
        discovery: backendPluginApi.coreServices.discovery,
        tokenManager: backendPluginApi.coreServices.tokenManager,
        scheduler: backendPluginApi.coreServices.scheduler,
        catalog: alpha.catalogServiceRef,
        indexRegistry: alpha$1.searchIndexRegistryExtensionPoint
      },
      async init({
        config,
        logger,
        discovery,
        tokenManager,
        scheduler,
        catalog,
        indexRegistry
      }) {
        const defaultSchedule = {
          frequency: { minutes: 10 },
          timeout: { minutes: 15 },
          initialDelay: { seconds: 3 }
        };
        const schedule = config.has("search.collators.techdocs.schedule") ? backendTasks.readTaskScheduleDefinitionFromConfig(
          config.getConfig("search.collators.techdocs.schedule")
        ) : defaultSchedule;
        indexRegistry.addCollator({
          schedule: scheduler.createScheduledTaskRunner(schedule),
          factory: pluginSearchBackendModuleTechdocs.DefaultTechDocsCollatorFactory.fromConfig(config, {
            discovery,
            tokenManager,
            logger: backendCommon.loggerToWinstonLogger(logger),
            catalogClient: catalog
          })
        });
      }
    });
  }
});

exports.searchModuleTechDocsCollator = searchModuleTechDocsCollator;
