// @backstage/plugin-scaffolder/dist/index.esm.js
import { r as rootRouteRef$1 } from './esm/plugin-34a96253.esm.js';
export { a as EntityNamePickerFieldExtension, E as EntityPickerFieldExtension, e as EntityPickerFieldSchema, b as EntityTagsPickerFieldExtension, j as EntityTagsPickerFieldSchema, M as MyGroupsPickerFieldExtension, l as MyGroupsPickerFieldSchema, k as MyGroupsPickerSchema, c as OwnedEntityPickerFieldExtension, i as OwnedEntityPickerFieldSchema, O as OwnerPickerFieldExtension, f as OwnerPickerFieldSchema, R as RepoUrlPickerFieldExtension, g as RepoUrlPickerFieldSchema, S as ScaffolderClient, d as ScaffolderPage, m as makeFieldSchemaFromZod, h as repoPickerValidation, s as scaffolderPlugin } from './esm/plugin-34a96253.esm.js';
export { T as TemplateTypePicker } from './esm/TemplateTypePicker-efec8c34.esm.js';
export { T as TaskPage } from './esm/TaskPage-1a149db7.esm.js';
import { createScaffolderFieldExtension as createScaffolderFieldExtension$1, ScaffolderFieldExtensions as ScaffolderFieldExtensions$1, useTemplateSecrets as useTemplateSecrets$1, scaffolderApiRef as scaffolderApiRef$1, createScaffolderLayout as createScaffolderLayout$1, ScaffolderLayouts as ScaffolderLayouts$1 } from '@backstage/plugin-scaffolder-react';
import '@backstage/integration-react';
import '@backstage/catalog-model';
import '@backstage/errors';
import 'qs';
import 'zen-observable';
import 'event-source-polyfill';
import '@backstage/catalog-client';
import '@backstage/core-plugin-api';
import '@backstage/plugin-catalog-react';
import '@material-ui/core';
import '@material-ui/core/FormControl';
import '@material-ui/lab/Autocomplete';
import 'react';
import 'react-use/lib/useAsync';
import 'zod';
import 'zod-to-json-schema';
import '@material-ui/core/FormHelperText';
import '@material-ui/core/Input';
import '@material-ui/core/InputLabel';
import '@backstage/core-components';
import 'react-use/lib/useDebounce';
import 'react-use/lib/useEffectOnce';
import '@material-ui/lab';
import 'lodash/capitalize';
import '@material-ui/icons/CheckBox';
import '@material-ui/icons/CheckBoxOutlineBlank';
import '@material-ui/icons/ExpandMore';
import '@material-ui/core/Grid';
import '@material-ui/core/Step';
import '@material-ui/core/StepLabel';
import '@material-ui/core/Stepper';
import '@material-ui/core/styles';
import '@material-ui/core/Typography';
import '@material-ui/icons/Cancel';
import '@material-ui/icons/Check';
import '@material-ui/icons/FiberManualRecord';
import 'classnames';
import 'luxon';
import 'react-router-dom';
import 'react-use/lib/useInterval';
import '@material-ui/icons/Language';

const rootRouteRef = rootRouteRef$1;
const createScaffolderFieldExtension = createScaffolderFieldExtension$1;
const ScaffolderFieldExtensions = ScaffolderFieldExtensions$1;
const useTemplateSecrets = useTemplateSecrets$1;
const scaffolderApiRef = scaffolderApiRef$1;
const createScaffolderLayout = createScaffolderLayout$1;
const ScaffolderLayouts = ScaffolderLayouts$1;

export { ScaffolderFieldExtensions, ScaffolderLayouts, createScaffolderFieldExtension, createScaffolderLayout, rootRouteRef, scaffolderApiRef, useTemplateSecrets };
